### PR TITLE
Contracts & Harnesses for `unchecked_add`

### PR DESCRIPTION
**Pull request by Yenyun035**
Towards #59 

### Changes
* Added contracts for `unchecked_add` (located in `library/core/src/num/int_macros.rs` and `uint_macros.rs`)
* Added a harness for `unchecked_add` of each integer type
  * `i8`, `i16`, `i32`, `i64`, `i128`, `isize`, `u8`, `u16`, `u32`, `u64`, `u128`, `usize` --- 12 harnesses in total.

### Revalidation
1. Per the discussion in #59, we have to **build and run Kani from `feature/verify-rust-std` branch**.
2. To revalidate the verification results, run the following command. `<harness_to_run>` can be either `num::verify` to run all harnesses or `num::verify::<harness_name>` (e.g. `check_unchecked_add_i8`) to run a specific harness.
```
kani verify-std  "path/to/library" \
    --harness <harness_to_run> \
    -Z unstable-options \
    -Z function-contracts \
    -Z mem-predicates
```

Except `isize` and `usize` harnesses passing 1203 checks, all harnesses should pass the same 1229 checks (1200 checks on 09/24/2024):
```
SUMMARY:
 ** 0 of 1229 failed

VERIFICATION:- SUCCESSFUL
Verification Time: 2.4755886s

Complete - 10 successfully verified harnesses, 0 failures, 10 total.
```

### Sidenotes
~~I ran `rustfmt` on `mod.rs`, `int_macros.rs`, and `uint_macros.rs`, so there might be format changes to existing code.~~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/num/int_macros.rs
Comment (celinval) [inline library/core/src/num/int_macros.rs:515]: I don't think you need an ensures here. This ensure clause is trivially true since all possible values are between MIN and MAX.

File: library/core/src/num/mod.rs
Comment (celinval) [inline library/core/src/num/mod.rs:1407]: can you please revert this format change? Thanks!

File: library/core/src/num/uint_macros.rs
Comment (celinval) [inline library/core/src/num/uint_macros.rs:562]: same here. This is redundant.

File: library/core/src/num/mod.rs
Comment (celinval) [inline library/core/src/num/mod.rs:1599]: Can you also include `isize` and `usize`?

**Review by celinval — APPROVED** [review]
That's awesome! Thanks

Have you tried using a `proc_macro` for generating the harness? I think it will be cleaner, but not a requirement.

File: library/core/src/num/mod.rs
Comment (Yenyun035) [reply library/core/src/num/mod.rs:1599]: For sure. I'll add it.

File: library/core/src/num/int_macros.rs
Comment (Yenyun035) [reply library/core/src/num/int_macros.rs:515]: That makes sense. I'll remove ensures in both files.

**Comment by Yenyun035** [comment]
I haven't tried with`proc_macro`. I'll look into it. Thank you for reviewing this PR :)

**Comment by celinval** [comment]
> I haven't tried with`proc_macro`. I'll look into it. Thank you for reviewing this PR :)

Sorry, my bad. I meant just regular macros

**Comment by Yenyun035** [comment]
> Sorry, my bad. I meant just regular macros

Do you mean "[declarative macros](https://doc.rust-lang.org/book/ch19-06-macros.html#declarative-macros-with-macro_rules-for-general-metaprogramming)"? How could it be used? I guess the macro can take the method to verify and the input type (e.g. i8, i16). I'm not familiar with it. I'll research it more.

File: library/core/src/num/mod.rs
Comment (feliperodri) [inline library/core/src/num/mod.rs:1590]: This code is quite repetitive. Could you try replace it with a macro? For example,

```rust
macro_rules! generate_unchecked_add_harness {
    ($type:ty) => {
        paste::paste! {
            #[kani::proof_for_contract(<$type>::unchecked_add)]
            pub fn [<check_unchecked_add_ $type>]() {
                let num1: $type = kani::any::<$type>();
                let num2: $type = kani::any::<$type>();

                unsafe {
                    num1.unchecked_add(num2);
                }
            }
        }
    };
}

// Use the macro to generate proof functions for all types.
generate_unchecked_add_harness!(i8);
generate_unchecked_add_harness!(i16);
generate_unchecked_add_harness!(i32);
generate_unchecked_add_harness!(i64);
generate_unchecked_add_harness!(i128);
generate_unchecked_add_harness!(isize);
generate_unchecked_add_harness!(u8);
generate_unchecked_add_harness!(u16);
generate_unchecked_add_harness!(u32);
generate_unchecked_add_harness!(u64);
generate_unchecked_add_harness!(u128);
generate_unchecked_add_harness!(usize);
```

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: library/core/src/num/mod.rs
Comment (feliperodri) [inline library/core/src/num/mod.rs:1589]: You need to add
```rust
#[cfg(kani)]
```
to avoid CI failures.

File: library/core/src/num/mod.rs
Comment (Yenyun035) [reply library/core/src/num/mod.rs:1590]: Thank you! This is a better way. I'll modify it.

Comment (Yenyun035) [reply library/core/src/num/mod.rs:1590]: @feliperodri It looks like `paste` is a third-party crate. I  might need to add `paste = "1.0.15"` as a dependency to `Cargo.toml`. Am I allowed to do so? If so, where should I add this line? I tried `pub extern crate paste;` in `mod.rs` but it said "can't find crate".

Comment (celinval) [reply library/core/src/num/mod.rs:1590]: Please don't add dependencies to the standard library. You should be able to create a macro without the paste crate.

Comment (celinval) [reply library/core/src/num/mod.rs:1590]: For that, you need to add an extra argument to your harness that will be the name of the harness, and use that instead.

Comment (Yenyun035) [reply library/core/src/num/mod.rs:1590]: Thank you! I tested with the following code without using `paste`. I was able to run the harness :)

```
    macro_rules! generate_unchecked_math_harness {
        ($type:ty, $method:ident, $harness_name:ident) => {

            #[kani::proof_for_contract($type::$method)]
            pub fn $harness_name() {
                let num1: $type = kani::any::<$type>();
                let num2: $type = kani::any::<$type>();

                unsafe {
                    num1.$method(num2);
                }
            }
        }
    }

    generate_unchecked_math_harness!(i8, unchecked_add, checked_unchecked_add_i8);
```

Comment (Yenyun035) [reply library/core/src/num/mod.rs:1590]: @feliperodri  @celinval Our current approach is to create three macros:
* One for `add`, `sub`, `mul` - Require two integers for a single operation
* One for `shl` and `shr` - Require one integer and an u32 integer (bits to shift)
* One for `neg` - Require one integer

There might still be some repetitive code, but it is much less than before. I believe there is a better way to combine the three macros, but I have no idea how to do it at this moment. Does the current approach make sense?

**Review by feliperodri — APPROVED** [review]

File: library/core/src/num/mod.rs
Comment (rajathmCMU) [inline library/core/src/num/mod.rs:1704]: Hi @celinval @feliperodri , we would also need some advice on the limit range (0 u128 to 1_000_000_000_000_000 u128 or higher? ) , if we choose to consider this approach. Thanks!

**Comment by Yenyun035** [comment]
@feliperodri @celinval Hello! The previous build failed due to an unused import. I fixed that and all looks good now. Could you approve the workflow? Once it passes the checks, we'll get this PR merged. Thank you :)

**Review by feliperodri — APPROVED** [review]

**Comment by danielhumanmod** [comment]
Hi @Yenyun035, I am the member of AWS team 4, after our team merge the main branch with this commit, there is a compiling error happens, like:
```
error: Failed to resolve checking function i8 :: unchecked_add because Kani currently cannot resolve path including primitive types
    --> /Users/danieltu/Developer/verify-rust-std/library/core/src/num/mod.rs:1596:13
     |
1596 |             #[kani::proof_for_contract($type::$method)]
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
1645 |     generate_unchecked_math_harness!(i8, unchecked_add, checked_unchecked_add_i8);
     |     ----------------------------------------------------------------------------- in this macro invocation
     |
     = note: this error originates in the attribute macro `kani::proof_for_contract` which comes from the expansion of the macro `generate_unchecked_math_harness` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I was wondering if this changes using some custom version of Kani?

**Comment by QinyuanWu** [comment]
> Hi @Yenyun035, I am the member of AWS team 4, after our team merge the main branch with this commit, there is a compiling error happens, like:
> 
> ```
> error: Failed to resolve checking function i8 :: unchecked_add because Kani currently cannot resolve path including primitive types
>     --> /Users/danieltu/Developer/verify-rust-std/library/core/src/num/mod.rs:1596:13
>      |
> 1596 |             #[kani::proof_for_contract($type::$method)]
>      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ...
> 1645 |     generate_unchecked_math_harness!(i8, unchecked_add, checked_unchecked_add_i8);
>      |     ----------------------------------------------------------------------------- in this macro invocation
>      |
>      = note: this error originates in the attribute macro `kani::proof_for_contract` which comes from the expansion of the macro `generate_unchecked_math_harness` (in Nightly builds, run with -Z macro-backtrace for more info)
> ```
> 
> I was wondering if this changes using some custom version of Kani?

@zhassan-aws

**Comment by Yenyun035** [comment]
@QinyuanWu Hi. Regarding your question about kani, yes. In our case, we need to manually build kani from source to run proofs because the current kani release (installed via `cargo install`)  does not support resolving paths with primitive types. This was discussed in our Challenge Issue #59. Here are our steps to build kani from source:

[Build from source code](https://model-checking.github.io/kani/build-from-source.html#dependencies)
1. Clone the repo: git clone https://github.com/model-checking/kani.git
2. Cd into kani repo
3. `git submodule update --init`
4. Install dependencies: `./scripts/setup/macos/install_deps.sh`
    * If failed on XCode CLT: try remove and reinstall OR see system settings > software update 
    * If you want to Install dependencies **manually** on MacOS
         * (latest) `brew install cbmc`
         * (latest) [CBMC-viewer](https://github.com/model-checking/cbmc-viewer/releases/tag/viewer-3.9)
         * Kissat: Run `kani/scripts/setup/install_kissat.sh`
5. `source $HOME/.cargo/env`
6. `git checkout features/verify-rust-std`
7. `cargo build-dev`
8. For all Kani commands, run with `{path/to/kani/repo}/scripts/kani`

**Pull request merged**
